src/fs.rs

```diff
@@ -0,0 +1,189 @@
+use sys_traits::FsMetadata;
+use sys_traits::FsOpen;
+use sys_traits::FsRemoveFile;
+use sys_traits::FsRename;
+use sys_traits::OpenOptions;
+use sys_traits::SystemRandom;
+use sys_traits::ThreadSleep;
+
+use crate::get_atomic_path;
+use crate::normalize_path;
+
+/// Canonicalizes a path which might be non-existent by going up the
+/// ancestors until it finds a directory that exists, canonicalizes
+/// that path, then adds back the remaining path components.
+///
+/// Note: When using this, you should be aware that a symlink may
+/// subsequently be created along this path by some other code.
+pub fn canonicalize_path_maybe_not_exists(
+  sys: &impl FsCanonicalize,
+  path: &Path,
+) -> std::io::Result<PathBuf> {
+  let path = normalize_path(path);
+  let mut path = path.as_path();
+  let mut names_stack = Vec::new();
+  loop {
+    match sys.fs_canonicalize(path) {
+      Ok(mut canonicalized_path) => {
+        for name in names_stack.into_iter().rev() {
+          canonicalized_path = canonicalized_path.join(name);
+        }
+        return Ok(canonicalized_path);
+      }
+      Err(err) if err.kind() == ErrorKind::NotFound => {
+        names_stack.push(match path.file_name() {
+          Some(name) => name.to_owned(),
+          None => return Err(err),
+        });
+        path = match path.parent() {
+          Some(parent) => parent,
+          None => return Err(err),
+        };
+      }
+      Err(err) => return Err(err),
+    }
+  }
+}
+
+pub fn atomic_write_file_with_retries<
+  TSys: FsCreateDirAll
+    + FsMetadata
+    + FsOpen
+    + FsRemoveFile
+    + FsRename
+    + ThreadSleep
+    + SystemRandom,
+>(
+  sys: &TSys,
+  file_path: &Path,
+  data: &[u8],
+  mode: u32,
+) -> std::io::Result<()> {
+  let mut count = 0;
+  loop {
+    match atomic_write_file(sys, file_path, data, mode) {
+      Ok(()) => return Ok(()),
+      Err(err) => {
+        if count >= 5 {
+          // too many retries, return the error
+          return Err(err);
+        }
+        count += 1;
+        let sleep_ms = std::cmp::min(50, 10 * count);
+        sys.thread_sleep(std::time::Duration::from_millis(sleep_ms));
+      }
+    }
+  }
+}
+
+/// Writes the file to the file system at a temporary path, then
+/// renames it to the destination in a single sys call in order
+/// to never leave the file system in a corrupted state.
+///
+/// This also handles creating the directory if a NotFound error
+/// occurs.
+pub fn atomic_write_file<
+  TSys: FsCreateDirAll + FsMetadata + FsOpen + FsRemoveFile + FsRename + SystemRandom,
+>(
+  sys: &TSys,
+  file_path: &Path,
+  data: &[u8],
+  mode: u32,
+) -> std::io::Result<()> {
+  fn atomic_write_file_raw<TSys: FsOpen + FsRename + FsRemoveFile>(
+    sys: &TSys,
+    temp_file_path: &Path,
+    file_path: &Path,
+    data: &[u8],
+    mode: u32,
+  ) -> std::io::Result<()> {
+    let mut file = sys.fs_open(temp_file_path, &OpenOptions::write())?;
+    file.fs_file_set_permissions(mode)?;
```

Comment:
I copied this from the CLI, but I kind of feel like we should only be setting the file permissions on open (the OpenOptions allow for that). Thoughts?

Reply:
I'm not sure I understand what you mean by "setting the file permissions on open". It seems desirable that the file is written with proper permissions.

Reply:
@bartlomieju I did a comimt to show it.